commit c8f34f9f7813dae303198de2f84a0ac5bf0c530a
Author: Mark Robert Miller <markrmiller@apache.org>
Date:   Sat Dec 22 21:06:04 2012 +0000

    SOLR-3180: wait for leader to see our recovering state before peer sync as well
    
    git-svn-id: https://svn.apache.org/repos/asf/lucene/dev/trunk@1425342 13f79535-47bb-0310-9956-ffa450edef68

diff --git a/solr/core/src/java/org/apache/solr/cloud/RecoveryStrategy.java b/solr/core/src/java/org/apache/solr/cloud/RecoveryStrategy.java
index 3f5eec7..bed5a86 100644
--- a/solr/core/src/java/org/apache/solr/cloud/RecoveryStrategy.java
+++ b/solr/core/src/java/org/apache/solr/cloud/RecoveryStrategy.java
@@ -341,6 +341,18 @@ public class RecoveryStrategy extends Thread implements ClosableThread {
         }
         
         zkController.publish(core.getCoreDescriptor(), ZkStateReader.RECOVERING);
+        
+        
+        sendPrepRecoveryCmd(leaderBaseUrl, leaderCoreName);
+        
+        // we wait a bit so that any updates on the leader
+        // that started before they saw recovering state 
+        // are sure to have finished
+        try {
+          Thread.sleep(2000);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        }
 
         // first thing we just try to sync
         if (firstTime) {
@@ -388,17 +400,6 @@ public class RecoveryStrategy extends Thread implements ClosableThread {
 
         log.info("Starting Replication Recovery. core=" + coreName);
         
-        sendPrepRecoveryCmd(leaderBaseUrl, leaderCoreName);
-        
-        // we wait a bit so that any updates on the leader
-        // that started before they saw recovering state 
-        // are sure to have finished
-        try {
-          Thread.sleep(2000);
-        } catch (InterruptedException e) {
-          Thread.currentThread().interrupt();
-        }
-        
         log.info("Begin buffering updates. core=" + coreName);
         ulog.bufferUpdates();
         replayed = false;
